components/commands/common/accelerator_parsing.cc

```diff
@@ -0,0 +1,132 @@
+
+namespace commands {
+
+namespace {
+constexpr char kApplicationClose[] = "AppClose";
+constexpr char kApplicationNew[] = "AppNew";
+std::string KeyboardCodeToDomCodeString(ui::KeyboardCode code) {
+  if (code == ui::VKEY_CLOSE) {
+    return kApplicationClose;
+  }
+
+  if (code == ui::VKEY_NEW) {
+    return kApplicationNew;
+  }
+
+  auto domcode = ui::UsLayoutKeyboardCodeToDomCode(code);
+  return ui::KeycodeConverter::DomCodeToCodeString(domcode);
+}
+
+ui::KeyboardCode DomCodeStringToKeyboardCode(const std::string& key) {
+  if (key == kApplicationClose) {
+    return ui::VKEY_CLOSE;
+  }
+
+  if (key == kApplicationNew) {
+    return ui::VKEY_NEW;
+  }
+
+  auto domcode = ui::KeycodeConverter::CodeStringToDomCode(key);
+  return ui::DomCodeToUsLayoutKeyboardCode(domcode);
+}
+
+std::vector<std::string> GetModifierNames(ui::KeyEventFlags flags) {
+  std::vector<std::string> result;
+
+  if (flags & ui::EF_COMMAND_DOWN) {
+    result.push_back("Meta");
+  }
+
+  if (flags & ui::EF_CONTROL_DOWN) {
+    result.push_back("Control");
+  }
+
+  if (flags & ui::EF_ALT_DOWN) {
+    result.push_back("Alt");
+  }
+
+  if (flags & ui::EF_SHIFT_DOWN) {
+    result.push_back("Shift");
+  }
+
+  if (flags & ui::EF_FUNCTION_DOWN) {
+    result.push_back("Fn");
+  }
+
+  return result;
+}
+
+ui::KeyEventFlags GetModifierFromKeys(
+    const std::vector<std::string>& modifiers) {
+  ui::KeyEventFlags result = ui::EF_NONE;
+  if (base::Contains(modifiers, "Control")) {
+    result = result | ui::EF_CONTROL_DOWN;
+  }
+  if (base::Contains(modifiers, "Meta")) {
+    result = result | ui::EF_COMMAND_DOWN;
+  }
+  if (base::Contains(modifiers, "Alt")) {
+    result = result | ui::EF_ALT_DOWN;
+  }
+  if (base::Contains(modifiers, "Shift")) {
+    result = result | ui::EF_SHIFT_DOWN;
+  }
+  if (base::Contains(modifiers, "Fn")) {
+    result = result | ui::EF_FUNCTION_DOWN;
+  }
+  return result;
+}
+}  // namespace
+
+std::string ToKeysString(const ui::Accelerator& accelerator) {
+  std::vector<std::string> parts = GetModifierNames(accelerator.modifiers());
+  auto domcode = ui::UsLayoutKeyboardCodeToDomCode(accelerator.key_code());
+  ui::DomKey domkey;
+  ui::KeyboardCode out_code;
+  if (!ui::DomCodeToUsLayoutDomKey(domcode, 0, &domkey, &out_code)) {
+    parts.push_back("Unknown Key");
+  } else {
+    parts.push_back(ui::KeycodeConverter::DomKeyToKeyString(domkey));
+  }
+  return base::JoinString(parts, "+");
+}
+
+std::string ToCodesString(const ui::Accelerator& accelerator) {
+  auto parts = GetModifierNames(accelerator.modifiers());
+  parts.push_back(KeyboardCodeToDomCodeString(accelerator.key_code()));
+  return base::JoinString(parts, "+");
+}
+
+ui::Accelerator FromCodesString(const std::string& value) {
+  std::vector<std::string> parts = base::SplitString(
+      value, "+", base::TRIM_WHITESPACE, base::SPLIT_WANT_ALL);
+
+  auto keyname = parts[parts.size() - 1];
+  auto keycode = DomCodeStringToKeyboardCode(keyname.empty() ? "+" : keyname);
```

Comment:
Q. what will the |keycode| be when keyname.empty() ?

Reply:
Nice catch - this was left over from when I was using the `keyname` instead of the `keycode` because `+` is a valid `keyname` and it would get split by the `base::SplitString`. Now I'm using keycodes though so it can't happen. Removed.